### PR TITLE
Add is expression support

### LINKED ISSUE #22436: Add 'is' expression
**Description:**
Refer: https://ballerina.io/spec/lang/2020R1/#is-expr

### PR DESCRIPTION
**Pull request by lochana-chathura**
Fixes #22436 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 1]
```diff
@@ -2699,6 +2707,8 @@ private STNode parseExpressionRhs(OperatorPrecedence currentPrecedenceLevel, Syn
                 // that matches to the current operator precedence level, and continue.
                 SyntaxKind binaryOpKind = getOperatorKindToInsert(currentPrecedenceLevel);
                 return parseExpressionRhs(currentPrecedenceLevel, binaryOpKind, lhsExpr, isAssignmentLhs);
+            } else if (solution.ctx == ParserRuleContext.IS_KEYWORD) {
+                return parseExpressionRhs(currentPrecedenceLevel, SyntaxKind.IS_KEYWORD, lhsExpr, isAssignmentLhs);
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:2711]: Why do we need to special case is keyword here?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 2]
```diff
@@ -4702,5 +4716,37 @@ private boolean isUnaryOperator(SyntaxKind kind) {
...
+     * <code>
+     * is-expr := expression is type-descriptor
+     * </code>
+     *
+     * @param lhsExpr Preceding expression of the is expression
+     * @return Is expression node
+     */
+    private STNode parseIsExpression(STNode lhsExpr) {
+        STNode isKeyword = parseIsKeyword();
+        startContext(ParserRuleContext.IS_EXPRESSION);
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:4731]: Does these two lines needs to be switched?

Comment (SupunS) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:4731]: Do we need to start a context at all?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 3]
```diff
@@ -1421,7 +1423,12 @@ private STNode parseTypeDescriptor(SyntaxKind tokenKind) {
                 return parseNilTypeDescriptor();
             default:
                 STToken token = peek();
-                Solution solution = recover(token, ParserRuleContext.TYPE_DESCRIPTOR);
+                Solution solution;
+                if (token.kind == SyntaxKind.EOF_TOKEN) {
+                    solution = recover(token, ParserRuleContext.SIMPLE_TYPE_DESCRIPTOR);
+                } else {
+                    solution = recover(token, ParserRuleContext.TYPE_DESCRIPTOR);
+                }
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:1431]: ParserErrorHandle should handle this IMO

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 4]
```diff
@@ -824,6 +824,11 @@ private Result seekMatch(ParserRuleContext currentCtx, int lookahead, int curren
                 case UNARY_OPERATOR:
                     hasMatch = isUnaryOperator(nextToken);
                     break;
+                case IS_KEYWORD:
+                    hasMatch = nextToken.kind == SyntaxKind.IS_KEYWORD;
+                    break;
+                case IS_EXPRESSION:
+                    return seekInIsExpression(currentCtx, lookahead, currentDepth, matchingRulesCount);
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:831]: Can't we handle this in a similar way to the binary expression?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 5]
```diff
@@ -1520,6 +1527,9 @@ private ParserRuleContext getNextRule(ParserRuleContext currentCtx, int nextLook
                 return ParserRuleContext.UNARY_OPERATOR;
             case UNARY_OPERATOR:
                 return ParserRuleContext.EXPRESSION;
+            case IS_KEYWORD:
+                startContext(ParserRuleContext.IS_EXPRESSION);
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:1531]: startContext() is only called at one place currently. Can we stick to that?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 3 above)
Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:1431]: Without above modification `test("3 is", "is-expr/is_expr_assert_04.json");` fails.  In error recovery for EOF, we directly insert current ctx. For `TYPE_DESCRIPTOR` ctx, getExpectedTokenKind(currentCtx);` returns a `NONE` syntax kind. 
So I ended up getting a `java.lang.StackOverflowError` as `return parseTypeDescriptor(solution.tokenKind);` line calls repeatedly. `test("3 is ;", "is-expr/is_expr_assert_04.json");` Passes as there's a semicolon next to 'is' keyword

Comment (SupunS) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:1431]: I see..
What if we update the `getExpectedTokenKind()` method to return a `SIMPLE_TYPE`, for `TYPE_DESCRIPTOR` context?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 2 above)
Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:4731]: Without a starting context,  `getNextRuleForSimpleTypeDesc()` looks for a 'VARIABLE_NAME'. But for `is_expression` we need to have `expression_rhs` as the next rule right?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java (hunk 4 above)
Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:831]: that means we don't need `IS_EXPRESSION` in `ParserRuleContext[] EXPRESSIONS' array?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 2 above)
Comment (SupunS) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:4731]: ack.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 3 above)
Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:1431]: yeah. it works. let's do that way.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java (hunk 4 above)
Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:831]: I was mimicking `ACCESS_EXPRESSION` there btw

Comment (SupunS) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:831]: I see. Ok, let's keep this for now. 
I think we will have to refactor both of these to be aligned with the binary expression. Can you please create an issue?

Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:831]: sure.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 6]
```diff
@@ -1353,10 +1359,10 @@ private ParserRuleContext getNextRule(ParserRuleContext currentCtx, int nextLook
...
                 }
                 if (parentCtx == ParserRuleContext.NIL_TYPE_DESCRIPTOR) {
-                    endContext();
-                    // After parsing nil type descriptor all the other parsing is same as next rule of simple type
-                    return getNextRuleForTypeDescriptor();
-                }
+                endContext();
+                // After parsing nil type descriptor all the other parsing is same as next rule of simple type
+                return getNextRuleForTypeDescriptor();
+            }
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:1365]: Incorrect indentation. 
I'l merge this PR anyway. Please fix it in the next PR.

Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:1365]: sure

**Pull request merged**